core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -576,6 +576,62 @@ private[spark] class TaskSetManager(
   }
 
   /**
+   * Check whether the given task set has been blacklisted to the point that it can't run anywhere.
+   *
+   * It is possible that this taskset has become impossible to schedule *anywhere* due to the
+   * blacklist.  The most common scenario would be if there are fewer executors than
+   * spark.task.maxFailures. We need to detect this so we can fail the task set, otherwise the job
+   * will hang.
+   *
```

Comment:
yes, much better.  Also good point that its only O(maxTaskFailures), I hadn't realized that.